pkg/store/proxy.go

```diff
@@ -255,305 +270,209 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
+				frameTimeoutCtx, cancel := frameCtx(s.responseTimeout)
+				defer cancel()
+				var rr *recvResponse
+				select {
+				case <-seriesCtx.Done():
+					err := errors.Wrapf(seriesCtx.Err(), "failed to receive any data from %s", st.String())
 
-	currSeries *storepb.Series
-	recvCh     chan *storepb.Series
+					errMtx.Lock()
+					errs.Add(err)
+					errMtx.Unlock()
 
-	errMtx sync.Mutex
-	err    error
+					span.SetTag("err", err.Error())
+					span.Finish()
+					close(done)
+					return false
+				case <-frameTimeoutCtx.Done():
+					err := errors.Wrapf(frameTimeoutCtx.Err(), "failed to receive any data in %v from %s", s.responseTimeout, st.String())
 
-	name            string
-	partialResponse bool
+					errMtx.Lock()
+					errs.Add(err)
+					errMtx.Unlock()
 
-	responseTimeout time.Duration
-	closeSeries     context.CancelFunc
-}
+					span.SetTag("err", err.Error())
+					span.Finish()
+					close(done)
+					return false
+				case rr = <-rCh:
+				}
 
-type recvResponse struct {
-	r   *storepb.SeriesResponse
-	err error
-}
+				if rr.err == io.EOF {
+					span.Finish()
+					close(done)
+					return false
+				}
 
-func frameCtx(responseTimeout time.Duration) (context.Context, context.CancelFunc) {
-	frameTimeoutCtx := context.Background()
-	var cancel context.CancelFunc
-	if responseTimeout != 0 {
-		frameTimeoutCtx, cancel = context.WithTimeout(frameTimeoutCtx, responseTimeout)
-		return frameTimeoutCtx, cancel
-	}
-	return frameTimeoutCtx, func() {}
-}
+				if rr.err != nil {
+					err := errors.Wrapf(rr.err, "receive series from %s", st.String())
+					errMtx.Lock()
+					errs.Add(err)
+					errMtx.Unlock()
 
-func startStreamSeriesSet(
-	ctx context.Context,
-	logger log.Logger,
-	span tracing.Span,
-	closeSeries context.CancelFunc,
-	wg *sync.WaitGroup,
-	stream storepb.Store_SeriesClient,
-	warnCh directSender,
-	name string,
-	partialResponse bool,
-	responseTimeout time.Duration,
-	emptyStreamResponses prometheus.Counter,
-) *streamSeriesSet {
-	s := &streamSeriesSet{
-		ctx:             ctx,
-		logger:          logger,
-		closeSeries:     closeSeries,
-		stream:          stream,
-		warnCh:          warnCh,
-		recvCh:          make(chan *storepb.Series, 10),
-		name:            name,
-		partialResponse: partialResponse,
-		responseTimeout: responseTimeout,
-	}
+					span.SetTag("err", err.Error())
+					span.Finish()
+					close(done)
+					return false
+				}
 
-	wg.Add(1)
-	go func() {
-		seriesStats := &storepb.SeriesStatsCounter{}
-		bytesProcessed := 0
-
-		defer func() {
-			span.SetTag("processed.series", seriesStats.Series)
-			span.SetTag("processed.chunks", seriesStats.Chunks)
-			span.SetTag("processed.samples", seriesStats.Samples)
-			span.SetTag("processed.bytes", bytesProcessed)
-			span.Finish()
-			close(s.recvCh)
-			wg.Done()
-		}()
-
-		numResponses := 0
-		defer func() {
-			if numResponses == 0 {
-				emptyStreamResponses.Inc()
+				storeResponses[i].responses = append(storeResponses[i].responses, rr.r)
```

Comment:
Those slices will kill us, we might want to buffer / pool them at some point

Reply:
Yes but it's not trivial, we've tried to migrate away from gogoprotobuf but that work is still not finished/stuck. With gogoprotobuf, it's impossible to do pooling without writing some kind of custom plugin / or forking it. Because it's unmaintained, it's not worth investing more time into gogoprotobuf.